From 891bfd149b02b1219f9b363dbde9485da1037958 Mon Sep 17 00:00:00 2001
From: Peter Pentchev <roam@ringlet.net>
Date: Wed, 16 Jul 2008 12:34:39 +0300
Subject: Net::FTP: cope gracefully with a failed command

In the Net::FTP::_data_cmd() routine the dataconn() sub is invoked
unconditionally and without any error checking immediately after sending
a command to the server.  It turns out that there are FTP servers out
there on the Big Bad Internet that close the data connection port
immediately after a failed transfer command.  In this case, Net::FTP
misbehaves in two ways:

1. Net::FTP::dataconn->close() and _close() methods try to use the
'net_ftp_cmd' member without checking if it has even been set - and on a
failed connection, it is not, thus Perl 5.10.0 breaks out with an error.

2. Net::FTP::_data_cmd() invokes dataconn() unconditionally even when it
does not *need* it - even when the command has failed.

This patch adds a Net::FTP::dataconn check whether the 'net_ftp_cmd'
member is defined at all before using it, and modifies
Net::FTP::_data_cmd() to invoke dataconn() in an eval, so that a failure
there will *still* return properly if the result code is bad.

Bug-Debian: http://bugs.debian.org/491062
Bug: https://rt.cpan.org/Public/Bug/Display.html?id=37700
Origin: https://rt.cpan.org/Public/Bug/Display.html?id=37700
Patch-Name: fixes/net_ftp_failed_command.diff
---
 cpan/libnet/Net/FTP.pm          |    4 +++-
 cpan/libnet/Net/FTP/dataconn.pm |    3 ++-
 2 files changed, 5 insertions(+), 2 deletions(-)

diff --git a/cpan/libnet/Net/FTP.pm b/cpan/libnet/Net/FTP.pm
index 9ed6d38..51efaae 100644
--- a/cpan/libnet/Net/FTP.pm
+++ b/cpan/libnet/Net/FTP.pm
@@ -1011,7 +1011,9 @@ sub _data_cmd {
 
     if ($ok) {
       $ftp->command($cmd, @_);
-      $data = $ftp->_dataconn();
+      eval {
+	$data = $ftp->_dataconn();
+      };
       $ok   = CMD_INFO == $ftp->response();
       if ($ok) {
         $data->reading
diff --git a/cpan/libnet/Net/FTP/dataconn.pm b/cpan/libnet/Net/FTP/dataconn.pm
index e7645cb..7b746ae 100644
--- a/cpan/libnet/Net/FTP/dataconn.pm
+++ b/cpan/libnet/Net/FTP/dataconn.pm
@@ -52,7 +52,7 @@ sub _close {
   $data->SUPER::close();
 
   delete ${*$ftp}{'net_ftp_dataconn'}
-    if exists ${*$ftp}{'net_ftp_dataconn'}
+    if defined($ftp) && exists ${*$ftp}{'net_ftp_dataconn'}
     && $data == ${*$ftp}{'net_ftp_dataconn'};
 }
 
@@ -69,6 +69,7 @@ sub close {
 
   $data->_close;
 
+  return 0 unless defined($ftp);
   $ftp->response() == CMD_OK
     && $ftp->message =~ /unique file name:\s*(\S*)\s*\)/
     && (${*$ftp}{'net_ftp_unique'} = $1);
